contents/javascript/Javascript-framework-React.md

```diff
@@ -0,0 +1,152 @@
+
+#### Benefit: Reusability
+
+Suppose you have a web application with the following markup in its UI:
+
+```html
+<h3>This is line 1.</h3>
+<h3>This is line 2.</h3>
+<h3>This is line 3.</h3>
+```
+
+The only difference in the three headings above is the number at the end of each line. Wouldn't it be nice if we could reuse a single line of markup for each heading and just pass in a different number each time?
+
+We can do this with React! Let's see how:
+
+<iframe src="https://codesandbox.io/embed/884m88wr90?fontsize=12" style="width:100%; height:400px; border:0; border-radius: 4px; overflow:hidden;" sandbox="allow-modals allow-forms allow-popups allow-scripts allow-same-origin"></iframe>
+
+<box type="info">
+    Notice the HTML-like syntax within the JavaScript method <code>render()</code>. This is <a href="https://reactjs.org/docs/introducing-jsx.html">JSX</a>, and it is a JavaScript syntax extension developed by the React team. React <a href="https://reactjs.org/docs/introducing-jsx.html#why-jsx">recommends</a> developers to use JSX, but it is optional to use it.
+</box>
+
+In the example above, we reused the markup for each heading using an `ExampleComponent`. A React **component** is just a JavaScript function that accepts arbitrary input (called `props`) and returns elements describing what should appear on the screen. 
+
+Components can be especially useful when your UI is complex. With components, you can not only *reuse* code, but also *divide* your UI into encapsulated units that manage their own data and logic. 
+
+For example, if you have a todo list application, it could have a `Form` component to handle adding new items and a `TodoList` component to display all the items. The `TodoList` component could further be composed of multiple `TodoListItem` components, one for each item in the todo list. 
+
+![Building a Todo List application using React](javascript-framework-react-imgs/ReactComponentsExample.png "Building a Todo List application using React")
+
+You can read more about components and props in the React [documentation](https://reactjs.org/docs/components-and-props.html). 
+
+#### Benefit: Better Performance
+
+Web applications can have a lot of user interaction and data updates, which results in changes being made to the DOM. Adding and removing DOM nodes isn't slow. The performance-bottleneck that arises due to DOM manipulation is because the browser needs to [reflow and repaint](https://www.phpied.com/rendering-repaint-reflowrelayout-restyle/) the UI every time the DOM is changed.
+
+React minimizes this update time by using a virtual DOM. The virtual DOM is a JavaScript object that is kept in the memory of your application.
+
+![React's Virtual DOM](javascript-framework-react-imgs/ReactVirtualDOM.png "React's Virtual DOM")
```

Comment:
`Actual DOM Updated`: I think a bit more info is needed to bring across the point that you're trying to make. Either in the diagram itself, a title for the diagram (eg. `Figure 2. ...`), or the paragraph below (I think the latter options are easier :P)